core/renamer.ml

```diff
@@ -54,11 +54,21 @@ let renaming_type_visitor instantiation_map =
 
     method! typ : datatype -> (datatype * 'self_type) = function
       | ForAll (qs, t) ->
-         let inst_map = shadow_vars inst_map qs in
-         let t', _ = {< inst_map >}#typ t in
-         ForAll (qs, t'), o
+         let qs', _ = List.split (List.map {< inst_map >}#quantifier qs) in
```

Comment:
This also seems wrong though: the problem isn't that we want to allow renaming *all* forall-quantified variables, it is that we want to rename *just* the ones in signatures that need to stay synchronized with those in the type annotation on the binder.

If you disable all shadowing then if the function signature is something like `forall a b. (forall c. a -> c) -> b` and the renamer applies a renaming such as `[a -> c, b -> d]` to the outer two quantifiers.  If shadowing is completely disabled then you get `forall c d. (forall c. c -> c) -> d` which is alpha-inequivalent.

This seems unlikely to actually happen provided the RHS of the renaming is all fresh variables, and maybe that is enough to ensure that this is correct, but it still feels a little wrong.  In that case, it would be a good idea to add a sanity check that whenever we encounter a variable to rename, that variable does *not* occur in the RHS of the renaming (that is, the freshly chosen variables were indeed fresh).  If we're generating fresh names this should always be true, but it would prevent inadvertently using the renamer unsoundly.

Reply:
> we want to rename just the ones in signatures that need to stay synchronized with those in the type annotation on the binder.

But we rename all variables in a signature, and all need to stay synchronized.

> If you disable all shadowing then (...)

Like you say, capture shouldn't be a problem if all variables in the codomain are fresh.

> it would be a good idea to add a sanity check that whenever we encounter a variable to rename, that variable does not occur in the RHS of the renaming

I'm hesitant to implement such a check for performance reasons and the fact that this seems rather unlikely to happen.

Reply:
OK.  Given that the external interface doesn't allow calling the renamer with arbitrary renamings this seems fine.